### PR TITLE
Design for IPA-to-IPA migration

### PR DESCRIPTION
**Pull request by rcritten**
There are some references and thoughts on expanding the
current migrate-ds plugin as well.

Related: https://pagure.io/freeipa/issue/3656
Signed-off-by: Rob Crittenden <rcritten@redhat.com>

### CODE REVIEW

**Comment by rcritten** [comment]
I'm concerned I've made certain assumptions that I just can't see because I'm too close to it.

File: doc/designs/migration.md [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+3. The quality of the data is unknown. 389-ds enforces proper syntax where other servers may not.
+
+### IPA to IPA migration
+
+For an IPA to IPA migration it is possible to make certain assumptions
+that aren't possible in a pure LDAP migration including:
+
+1. The DIT is understood and consistent
+2. Unique UID/GID/login/group name(s) so additional checking is not necessary
+3. A consistent schema (perhasp variability for users and groups)
```
Comment (flo-renaud) [inline doc/designs/migration.md:85]: type: perhasp

File: doc/designs/migration.md [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+user-modifiable and if a user decides to mess with them, they can correct it
+after migration.
+
+#### Migrating cn=config
+
+A number of plugins are optional in IPA: nis, compat, ACME.
+
+These will be enabled in the new IPA server if enabled in the remote:
+
+* cn=Schema Compatibility, cn=plugins, cn=config
```
Comment (flo-renaud) [inline doc/designs/migration.md:249]: Note: if the schema compat plugin is enabled, entries below cn=compat,$BASEDN must be skipped by the migration command. They would be dynamically re-created on the target LDAP server.

File: doc/designs/migration.md [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+https://pagure.io/freeipa/issue/3100 - Check for userPassword in migration
+https://pagure.io/freeipa/issue/4738 - [RFE] ipa migrate-ds should provide option for creating UPG from posixGroup objectClass
+https://pagure.io/freeipa/issue/5020 - migrate-ds: does not show migrated users if an error happened during group migration
+https://pagure.io/freeipa/issue/5693 - Passwords become "expired" when migrating from directory server to IPA
+https://pagure.io/freeipa/issue/6105 - migrate-ds is not completely ignoring attributes.
+https://pagure.io/freeipa/issue/6360 - ipa migrate-ds does not rename uniquemember/member attributes properly
+https://pagure.io/freeipa/issue/6380 - ipa migrate-ds should print warning for referrals
+https://pagure.io/freeipa/issue/7368 - ipa migrate-ds converts groupofuniquenames objects to groupofnames, but leaves groupofuniquenames objectclass present
+https://pagure.io/freeipa/issue/7749 - `ipa migrate-ds` fails to migrate user and group data from directory server to IDM.
+
```
Comment (flo-renaud) [inline doc/designs/migration.md:367]: Formatting: the above lines are concatenated into a single one. You can mark each one as a list element for a nicer output.

Comment (rcritten) [reply doc/designs/migration.md]: Comments addressed.

I also renamed the design to ipa_to_ipa_migration.md to be a bit more specific.

**Comment by abbra** [comment]
@rcritten could you please rebase it onto master to allow ReadTheDocs to pick up and build the design docs?

**Comment by abbra** [comment]
We can merge this and improve when actual IPA-IPA migration tool is written.

**Comment by abbra** [comment]
master:

* d4859db4ef775f1ac06f768dcc53daeb315b1c96 Design for IPA-to-IPA migration

**Pull request closed without merging**